pennylane_lightning/core/src/simulators/lightning_qubit/catalyst/StateVectorLQubitDynamic.hpp

```diff
@@ -0,0 +1,387 @@
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#pragma once
+
+#include <numeric>
+#include <utility>
+#include <vector>
+
+#include "BitUtil.hpp"
+#include "LinearAlgebra.hpp"
+#include "Util.hpp"
+
+#include "Error.hpp"
+
+#include <StateVectorLQubit.hpp>
+
+#include <iostream>
+
+namespace {
+using Pennylane::Util::AlignedAllocator;
+using Pennylane::Util::bestCPUMemoryModel;
+using Pennylane::Util::exp2;
+using Pennylane::Util::isPerfectPowerOf2;
+using Pennylane::Util::log2PerfectPower;
+using Pennylane::Util::ONE;
+using Pennylane::Util::squaredNorm;
+using Pennylane::Util::ZERO;
+} // namespace
+
+namespace Pennylane::LightningQubit {
+/**
+ * @brief State-vector dynamic class.
+ *
+ * This class allocates and deallocates qubits/wires dynamically,
+ * and defines all operations to manipulate the statevector data for
+ * quantum circuit simulation.
+ *
+ */
+template <class fp_t = double>
+class StateVectorLQubitDynamic
+    : public StateVectorLQubit<fp_t, StateVectorLQubitDynamic<fp_t>> {
+  public:
+    using PrecisionT = fp_t;
+    using ComplexT = std::complex<PrecisionT>;
+    using CFP_t = ComplexT;
+    using MemoryStorageT = Pennylane::Util::MemoryStorageLocation::Internal;
+
+  private:
+    using BaseType =
+        StateVectorLQubit<PrecisionT, StateVectorLQubitDynamic<PrecisionT>>;
+    std::vector<ComplexT, AlignedAllocator<ComplexT>> data_;
+
+    static constexpr PrecisionT epsilon_ =
+        std::numeric_limits<PrecisionT>::epsilon() * 100;
+
+    template <class IIter, class OIter>
+    inline OIter _move_data_elements(IIter first, size_t distance,
+                                     OIter second) {
+        *second++ = std::move(*first);
+        for (size_t i = 1; i < distance; i++) {
+            *second++ = std::move(*++first);
+        }
+        return second;
+    }
+
+    template <class IIter, class OIter>
+    inline OIter _shallow_move_data_elements(IIter first, size_t distance,
+                                             OIter second) {
+        for (size_t i = 0; i < distance; i++) {
+            *second++ = std::move(*first);
+            *first = ZERO<PrecisionT>();
+            first++;
+        }
+        return second;
+    }
+
+    inline void
+    _scalar_mul_data(std::vector<ComplexT, AlignedAllocator<ComplexT>> &data,
+                     ComplexT scalar) {
+        std::transform(
+            data.begin(), data.end(), data.begin(),
+            [scalar](const ComplexT &elem) { return elem * scalar; });
+    }
+
+    inline void
+    _normalize_data(std::vector<ComplexT, AlignedAllocator<ComplexT>> &data) {
+        _scalar_mul_data(data,
+                         ONE<PrecisionT>() /
```

Comment:
Same q as above?

Reply:
@rauletorresc There seems to be a few missing unit tests in this PR comparing to [Test_LightningDriver.cpp](https://github.com/PennyLaneAI/catalyst/blob/main/runtime/tests/Test_LightningDriver.cpp#L97-L291). These helper methods were tested implicitly there if not mistaken (at least they used to!)

Reply:
Fixed

Reply:
@maliasadi I have brought from Catalyst the driver tests and the SVD tests. Thanks to that, most of the code coverage complains have been eliminated, with the exception of this one. Do you have an idea which tests in Catalyst might remove this one as well?